src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,289 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | A 'Ledger' provides access to a blockchain as well as the state
+-- and receipt associated to a block in the blockchain.
+--
+-- This module should be imported qualified.
+module Oscoin.Storage.Ledger
+    ( Ledger
+    , hoist
+    , mkLedger
+    , newMemoryIO
+    , newMemoryFromGenesisIO
+    , newFromBlockStoreIO
+
+    , blockStore
+    , blockStoreReader
+    , getBlocksByDepth
+    , getTipWithState
+    , getTip
+    , lookupState
+    , lookupReceipt
+
+    , buildNextBlock
+    , addBlock
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain.Block
+import           Oscoin.Crypto.Blockchain.Eval
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Storage.Block.Abstract (BlockStore)
+import qualified Oscoin.Storage.Block.Abstract as BlockStore
+import           Oscoin.Storage.Block.Memory
+import           Oscoin.Storage.HashStore
+import qualified Oscoin.Storage.Receipt as ReceiptStore
+import qualified Oscoin.Time.Chrono as Chrono
+
+import           Codec.Serialise (Serialise)
+import           Control.Monad.Trans.Maybe
+import qualified Crypto.Data.Auth.Tree.Class as AuthTree
+import qualified Data.List.NonEmpty as NonEmpty
+
+-- | Errors that may be raised from the ledger's methods
+data LedgerError c
+    = StateNotInCache (BlockHash c)
+    -- ^ State for the given block is not materialized in the state
+    -- cache.
+
+deriving instance (Crypto.HasHashing c) => Eq (LedgerError c)
+deriving instance (Show (Crypto.Hash c)) => Show (LedgerError c)
+
+data Ledger crypto seal tx output state (m :: * -> *) = Ledger
+    { ledgerBlockStore   :: BlockStore crypto tx seal m
```

Comment:
I have discussed this with @geigerzaehler in a call and, in my humble opinion, this is the wrong design move. The `Ledger` is about the computation/application level, whereas the `Protocol` and the `Blockstore` deals exclusively with the settlement layer.

Therefore, it feels iffy to me that we are now shunting the operations on the main chain (i.e. `addBlock` and whatnot) to a component of the system what shouldn't worry about it.

Rather, I would like to give the `Ledger` only a `BlockStoreReader` handle in case it needs to read the state of the main (adopted) chain, but keep both the `Ledger` and the `BlockStore` as part of the `Protocol`. We don't lose anything by doing so: the `Protocol` is still the only owner of the `BlockStoreWriter`, whereas it feels totally normal to have multiple readers for the chain.

I would be curious to hear what folks think about this.

Reply:
Let’s bring this up during the sync and find a place to discuss this

Reply:
Sounds like a reasonable concern. It would mean that `stepBlock` is responsible for validation, `materializeBlock`-ing, and storing the block, is that correct?

Reply:
If by `stepBlock` you mean `stepProtocol` (sorry, I am starting to become carb-depleted 😁 ) I think this would align with my view that the `Procotol` was this overarching component which would perform chain-selection and make sure our view of the chain would be consistent. Originally I did envision a schema where we would simply "push out" events to consumers (like the `Ledger` for example) in the style of `[AddBlock, RollbackBlock,...]` and that the `Ledger` would have been able to update its state accordingly. However , apparently in our current system we cannot do that directly as a block must be evaluated first before it can be considered valid for inclusion. This means that yes, we do need the `Ledger` in scope in the `Protocol` to perform such step (which is unfortunate, but it is the way it is) but we can potentially still retain full control over the chain and its consistency. One very strong argument would also be that `stepProtocol` runs in a critical section and we need to ensure that:

1. We keep the lock as little as possible;
2. We do not call into slow sub-routines which might delay block applications.

(This is slightly orthogonal to the concern discussed here but I guess it's good to remember while we discuss the design).

Reply:
> I have discussed this with @geigerzaehler in a call and, in my humble opinion, this is the wrong design move. The `Ledger` is about the computation/application level, whereas the `Protocol` and the `Blockstore` deal exclusively with the settlement layer.
> 
Could this be in any way an argument against naming this "Ledger"? vs. against the chosen abstraction?

Reply:
> Could this be in any way an argument against naming this "Ledger"? vs. against the chosen abstraction?

Uhm, maybe (thinking out loud here as always): I guess the sour taste in my mouth comes from the fact that, in an ideal world, the `Ledger` would only deal with materialising the state by being fed with information coming from the settlement/consensus layer (as in: this block has been added, these many blocks have been rolled back etc) and at the moment we cannot achieve such clean split due to this validation/evaluation business, which perhaps calls for another component/different design somehow?